torch/distributed/pipelining/schedules.py

```diff
@@ -656,12 +663,19 @@ def _step_microbatches(
         stage_index_to_stage: Dict[int, _PipelineStageBase] = {
             stage.stage_index: stage for stage in self._stages
         }
-        prev_rank: int = (self.rank - 1) % self.pp_group_size
-        next_rank: int = (self.rank + 1) % self.pp_group_size
+
+        # determine prev_rank and next_rank based on which ranks are next to
+        # the stages in the pipeline_order
+        all_prev_ranks: Set[int] = set()
```

Comment:
the naming (all) suggests more than one next rank, which is a little confusing.

i think the intent is that later we could add skip connection support?
- if we added it, i'd want to rename perhaps to 'fwd_send_ranks' rather than using the word 'next' which implies linearity together with 'all' which applies multiple

Reply:
This will also help to support skip connections. But this is still needed if we truly want to support arbitrary stage per rank since there could be multiple prev ranks. For example:

rank0: [stage0, stage2]
rank1: [stage1, stage4]
rank2: [stage2, stage3]

rank1 have a previous rank0 (from stage0) and previous rank2 (from stage3). This is a contrived example, but this list containing multiple ranks is needed if we want to have no restrictions for the user.